flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/index/IndexGeneratorFactory.java

```diff
@@ -0,0 +1,294 @@
+					@Override
+					public String generate(Row row) {
+						LocalDateTime indexField = (LocalDateTime) row.getField(indexFieldPos);
+						String indexFieldValueStr = indexField.format(dateTimeFormatter);
+						return indexPrefix.concat(indexFieldValueStr).concat(indexSuffix);
+					}
+				};
+			}
+			else if (indexFieldType == Types.SQL_TIMESTAMP) {
+				return new AbstractTimeIndexGenerator(dateTimeFormat) {
+					@Override
+					public String generate(Row row) {
+						Timestamp indexField = (Timestamp) row.getField(indexFieldPos);
+						String indexFieldValueStr = indexField.toLocalDateTime().format(dateTimeFormatter);
+						return indexPrefix.concat(indexFieldValueStr).concat(indexSuffix);
+					}
+				};
+			}
+			// DataTypes.SQL_DATE
+			else if (indexFieldType == Types.LOCAL_DATE) {
+				return new AbstractTimeIndexGenerator(dateTimeFormat) {
+					@Override
+					public String generate(Row row) {
+						LocalDate indexField = (LocalDate) row.getField(indexFieldPos);
+						String indexFieldValueStr = indexField.format(dateTimeFormatter);
+						return indexPrefix.concat(indexFieldValueStr).concat(indexSuffix);
+					}
+				};
+			} else if (indexFieldType == Types.SQL_DATE) {
+				return new AbstractTimeIndexGenerator(dateTimeFormat) {
+					@Override
+					public String generate(Row row) {
+						Date indexField = (Date) row.getField(indexFieldPos);
+						String indexFieldValueStr = indexField.toLocalDate().format(dateTimeFormatter);
+						return indexPrefix.concat(indexFieldValueStr).concat(indexSuffix);
+					}
+				};
+			} // DataTypes.TIME
+			else if (indexFieldType == Types.LOCAL_TIME) {
+				return new AbstractTimeIndexGenerator(dateTimeFormat) {
+					@Override
+					public String generate(Row row) {
+						LocalTime indexField = (LocalTime) row.getField(indexFieldPos);
+						String indexFieldValueStr = indexField.format(dateTimeFormatter);
+						return indexPrefix.concat(indexFieldValueStr).concat(indexSuffix);
+					}
+				};
+			} else if (indexFieldType == Types.SQL_TIME) {
+				return new AbstractTimeIndexGenerator(dateTimeFormat) {
+					@Override
+					public String generate(Row row) {
+						Time indexField = (Time) row.getField(indexFieldPos);
+						String indexFieldValueStr = indexField.toLocalTime().format(dateTimeFormatter);
+						return indexPrefix.concat(indexFieldValueStr).concat(indexSuffix);
+					}
+				};
+			} else {
+				throw new TableException(String.format("Unsupported type '%s' found in Elasticsearch dynamic index field, " +
+					"time-related pattern only support types are: %s", indexFieldType, indexHelper.getSupportedDateTypes()));
+			}
+		}
+		// general dynamic index pattern
+		return new DefaultIndexGenerator(index) {
+
+			@Override
+			public String generate(Row row) {
+				Object indexField = row.getField(indexFieldPos);
+				return indexPrefix.concat(indexField == null ? index : indexField.toString()).concat(indexSuffix);
```

Comment:
Why use `index` when indexField is null? For example, `prefix_{my_field}_index`, if `my_field` is null, the generated index will be `prefix_prefix_{my_field}_index_index` which is confusing. 

Shound't we just drop this record if the field is null?

Reply:
prefix_null_index